### PR TITLE
[WIP] A framework for distributed Word2Vec using Distributed Tensorflow

### PR DESCRIPTION
**Pull request by souravsingh**
@tmylk  This is a work in progress feature to add support for Distributed Word2Vec in Gensim using Distributed Tensorflow. 

Apache Spark has some complications with setting up worker clusters for the jobs, hence the reason to use Tensorflow.

### CODE REVIEW

**Comment by tmylk** [comment]
Thanks! Do you have a benchmark similar to #1033?

**Comment by souravsingh** [comment]
The code is currently not tested for benchmarks. I will start working on it.

**Comment by tmylk** [comment]
What is the advatage of this code over #1033  by @markroxor and @phalexo?

**Comment by souravsingh** [comment]
This can make use of GPUs over multiple slave clusters if it is properly configured

**Comment by tmylk** [comment]
@souravsingh Please post the latest benchmark discussed on gitter

**Comment by souravsingh** [comment]
Here are the observations-

Word2Vec run on a 4-core machine in Google cloud with Batch Size=128, Embedding Size=128, Steps for Training=100001, neg = 0-
`` Time= 246 seconds, Loss rate=4.71``

Word2Vec run on a two 4-core machines in Google cloud with all parameters same as above and one machine acting as parameter server and another acting as worker-
`` Time= 460 seconds, Loss rate=4.89``

Word2Vec run on a three 4-core machines in Google cloud with all parameters same as above and one machine acting as parameter server and two machines acting as worker-
`` Time= 190 seconds, Loss rate=4.68``

**Comment by souravsingh** [comment]
@tmylk Is there anything else that needs to be done?

**Comment by phalexo** [comment]
I am not sure how various parameters settings compare to the ones' I was asked to use but my recollection is that in some benchmarks I got running word2vec on 2 XEON (8-cores each) machine were under 10 seconds for a single epoch. Was I using completely different parameters? The vector I used was 100, but that is a minor difference.  How do the numbers below compare?

What is "steps?" Number of epochs?

"""
Word2Vec run on a 4-core machine in Google cloud with Batch Size=128, Embedding Size=128, Steps for Training=100001, neg = 0-
Time= 246 seconds, Loss rate=4.71

Word2Vec run on a two 4-core machines in Google cloud with all parameters same as above and one machine acting as parameter server and another acting as worker-
Time= 462 seconds, Loss rate=4.89

Word2Vec run on a three 4-core machines in Google cloud with all parameters same as above and one machine acting as parameter server and two machines acting as worker-
Time= 198 seconds, Loss rate=4.78
"""

**Comment by souravsingh** [comment]
@phalexo The steps are number of epochs in this case. I cannot get access to 8-core machine in Google Cloud due to restrictions in free trial phase.

**Comment by phalexo** [comment]
@souravsingh I'd be curious to see the exact parameters you are passing to word2vec and also about which input file is being used, to compare with what I did. Thx.

**Comment by souravsingh** [comment]
I made use of the text8 corpus for the input. And these are the parameters for training-
Embedding Size=128
Number of Negative Samples=64
Batch Size=128
Window Size=1
Subsampling rate=1e-2
Vocabulary size = 50000
Number of Steps= 100001
Epochs= 500

**Comment by phalexo** [comment]
@souravsingh So, the number of epochs is 500, not 100001. What is the "100001" then? 
window=1 would not make any sense to me to benchmark. The whole point of word2vec is to place words in relationship with other words that surround it. Unless I misunderstand what you mean by window=1, a window of 1 word would not make sense. As window size grows so does the time.

**Comment by souravsingh** [comment]
It is the number of steps for which the Word2Vec is trained, which is 100001.

There is a parameter for called num_skips, which is 2, which tells how many times to reuse an input to generate a label.

**Comment by tmylk** [comment]
@souravsingh For completeness, could you please paste the code for benchmark?

**Comment by souravsingh** [comment]
I adapted the code from the example here- https://github.com/tensorflow/tensorflow/blob/master/tensorflow/examples/tutorials/word2vec/word2vec_basic.py for running on single machine.

**Comment by menshikh-iv** [comment]
Ping @souravsingh, what status of this PR?

**Comment by souravsingh** [comment]
@menshikh-iv I am currently facing an issue with creation of servers through the script.

**Comment by menshikh-iv** [comment]
@souravsingh You can use [Ansible](https://www.ansible.com/) for this purpose. Let us know if we can help you.

**Comment by menshikh-iv** [comment]
Continues in #1479
